craft_store/creds.py

```diff
@@ -0,0 +1,153 @@
+"""Functions to serialize/deserialize credentials for Candid and Ubuntu One SSO."""
+
+import json
+from typing import Literal
+
+import pydantic
+from pydantic import BaseModel, Field
+
+from . import errors
+
+
+class CandidModel(BaseModel):
+    """Model for Candid credentials."""
+
+    t: Literal["macaroon"] = Field("macaroon")
+    v: str
+
+    @property
+    def value(self) -> str:
+        """Get the actual credentials string."""
+        return self.v
+
+    def marshal(self) -> str:
+        """Serialize this Candid model into a string suitable for storage."""
+        return json.dumps(self.dict())
+
+    @classmethod
+    def unmarshal(cls, marshalled_creds: str) -> "CandidModel":
+        """Deserialize a Candid model from previously-stored credentials.
+
+        :param marshalled_creds:
+            The previously-stored Candid credentials, which can either be in the "new",
+            type-based format or in the "old", raw-string one.
+        """
+        try:
+            creds = json.loads(marshalled_creds)
+        except json.JSONDecodeError:
+            return CandidModel(v=marshalled_creds)  # pyright: ignore
```

Comment:
Pydantic has a `Json` type that unmarshals json automatically, so you can have `CandidModel` as with `v: pydantic.Json` and `OldCandidModel` with `v: str` and no explicit unmarshaling logic.

```
class Foo(pydantic.BaseModel):
    x: Literal["foo"] = "foo"
    y: pydantic.Json

a = Foo(x="foo", y='{"a":42}')
print(a)
```
results in `x='foo' y={'a': 42}`